util/ValueSetHelper.ts

```diff
@@ -0,0 +1,179 @@
+import { TestCaseInfo } from '../state/atoms/patientTestCase';
+import fhirpath from 'fhirpath';
+import { parsedCodePaths } from '../util/codePaths';
+import { DataRequirementsLookupByTypeProps } from '../state/selectors/dataRequirementsLookupByType';
+import { CodeableConcept } from 'fhir/r2';
+
+export interface GetValueSetCodesProps {
+  code?: string;
+  system?: string;
+}
+
+/**
+ * Helper function that takes a ValueSet url and a measure bundle and returns an
+ * array of objects that contain each code and system in the ValueSet
+ */
+function getValueSetCodes(valueSetUrl: string[], mb: fhir4.Bundle | null): GetValueSetCodesProps[] {
+  const codesAndSystems: GetValueSetCodesProps[] = [];
+  valueSetUrl.forEach(vs => {
+    const vsResource = mb?.entry?.filter(r => r.resource?.resourceType === 'ValueSet' && r.resource?.url === vs)[0]
+      .resource as fhir4.ValueSet;
+    vsResource.expansion?.contains?.forEach(c => {
+      codesAndSystems.push({ code: c.code, system: c.system });
+    });
+  });
+  return codesAndSystems;
+}
+
+/**
+ * Helper function for repeated logic in minimizeTestCaseResources that takes a primaryCodeValue,
+ * a resource, a matching DataRequirement of the same resourceType, a measureBundle, and a newResources
+ * array, checks that the primaryCodeValue exists and if the primaryCodeValue matches any of the
+ * directCodes or ValueSet codes of a DataRequirement of the same resourceType, the resource is added
+ * to the newResources array
+ */
+function checkCodesAndValueSets(
+  primaryCodeValue: fhir4.CodeableConcept,
+  matchingDRType: DataRequirementsLookupByTypeProps,
+  measureBundle: fhir4.Bundle | null,
+  resource: fhir4.BundleEntry,
+  newResources: fhir4.BundleEntry[]
+) {
+  if (primaryCodeValue) {
+    if (
+      matchingDRType.directCodes.length > 0 &&
+      matchingDRType.directCodes.find(dc => primaryCodeValue.coding?.find(c => c.code === dc.code))
+    ) {
+      newResources.push(resource);
+    } else if (matchingDRType.valueSets.length > 0) {
+      const vsCodesAndSystems = getValueSetCodes(matchingDRType.valueSets, measureBundle);
+      if (
+        vsCodesAndSystems.find(vscas =>
+          primaryCodeValue.coding?.find(c => c.code === vscas.code && c.system === vscas.system)
+        )
+      ) {
+        newResources.push(resource);
+      }
+    }
+  }
+}
+
+/**
+ * Helper function that takes in a TestCase, a measure bundle, and the
+ * current measure bundle's dataRequirements lookup by type object and returns
+ * an array of BundleEntries that are TestCase resources that are relevant to the
+ * measure bundle's data requirements
+ */
+export function minimizeTestCaseResources(
+  testCase: TestCaseInfo,
+  measureBundle: fhir4.Bundle | null,
+  drLookupByType: Record<string, DataRequirementsLookupByTypeProps>
+): fhir4.BundleEntry[] {
+  const newResources: fhir4.BundleEntry[] = [];
+  testCase.resources.forEach(r => {
```

Comment:
There are multiple places that `newResources.push` is called here. It feels like this could have a chance for duplicated resources, likely when there are valueSets with overlapping codes. This could be changed to using `Array.filter` with this predicate function returning `true` when ever matching code is found. This would also allow for an easy short-circuit to help speed things up.

Reply:
I am actually not totally seeing how resources could be duplicated here, can you point to a spot in the logic where that could happen?